server/server.go

```diff
@@ -181,6 +187,32 @@ func (s *Server) Stop() {
 	s.stopper.Stop()
 }
 
+// ServeHTTP is necessary to implement the http.Handler interface. It
+// will snappy a response if the appropriate request headers are set.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if we're draining; if so return 503, service unavailable.
+	if !s.stopper.StartTask() {
+		http.Error(w, "service is draining", http.StatusServiceUnavailable)
+		return
+	}
+	defer s.stopper.FinishTask()
+
+	ae := r.Header.Get("Accept-Encoding")
```

Comment:
We should set `Vary: Accept-Encoding` here (or it's probably even better to disable caching completely with `Cache-control: no-cache`)